src/modules/job-manager/plugins/perilog.c

```diff
@@ -233,10 +236,121 @@ static void emit_finish_event (struct perilog_proc *proc,
+                    flux_log_error (h,
+                                    "failed to add rank=%lu to drain set",
+                                    rank);
+                }
+            }
+        }
+        rank = idset_next (proc->ranks, rank);
+    }
+    if (!(proc->failed_ranks = idset_encode (failed, IDSET_FLAG_RANGE))) {
+        flux_log_error (h,
+                        "%s: error encoding %s failed ranks",
+                        idf58 (proc->id),
+                        perilog_proc_name (proc));
+        goto out;
+    }
+    (void) snprintf (reason,
+                     sizeof (reason),
+                     "%s failed for job %s",
+                     perilog_proc_name (proc),
+                     idf58 (proc->id));
+
+    if (!(f = flux_rpc_pack (h,
+                             "resource.drain",
+                             0,
+                             0,
+                             "{s:s s:s s:s}",
+                             "targets", proc->failed_ranks,
+                             "reason", reason,
+                             "mode", "update"))) {
+        flux_log (h,
+                  LOG_ERR,
+                  "%s: %s: failed to send drain RPC for ranks %s",
+                  idf58 (proc->id),
+                  perilog_proc_name (proc),
+                  proc->failed_ranks);
+        goto out;
+    }
+out:
+    idset_destroy (failed);
+    return f;
+}
+
+
+static void drain_failed_cb (flux_future_t *f, void *arg)
+{
+    struct perilog_proc *proc = arg;
+    flux_t *h = flux_jobtap_get_flux (proc->p);
+
+    if (flux_future_get (f, NULL) < 0) {
+        flux_log (h,
+                  LOG_ERR,
+                  "Failed to drain ranks with failed %s for %s: %s",
+                  perilog_proc_name (proc),
+                  idf58 (proc->id),
+                  future_strerror (f, errno));
+    }
+    flux_future_destroy (f);
+    emit_finish_event (proc, proc->bulk_exec);
+    perilog_proc_delete (proc);
+}
+
+static bool perilog_per_rank (struct perilog_proc *proc)
+{
+    if (proc->prolog)
+        return perilog_config.prolog_per_rank;
+    return perilog_config.epilog_per_rank;
+}
+
 static void completion_cb (struct bulk_exec *bulk_exec, void *arg)
 {
     struct perilog_proc *proc = bulk_exec_aux_get (bulk_exec, "perilog_proc");
     if (proc) {
+        if (perilog_per_rank (proc)
+            && !proc->canceled
+            && bulk_exec_rc (bulk_exec) != 0) {
+            flux_future_t *f;
+
+            /* Drain the set of ranks that failed the prolog/epilog. If the
+             * drain RPC is successful, then wait for the response before
+             * emitting the "prolog/epilog-finish" event. O/w, resources could
+             * be freed and handed out to new jobs before they are drained.
+             */
+            if ((f = drain_failed_ranks (proc))
+                 && flux_future_then (f, -1., drain_failed_cb, proc) == 0)
+                return;
```

Comment:
If perilog is unloaded with this future pending, seems like the job manager reactor could end up calling a bad address?

Should the future be part of the proc struct so it can be destroyed if it is in progress when perilog is unloaded?

Reply:
Yeah, good catch! Note that unloading the perilog plugin with outstanding prolog/epilog will result in jobs with `{prolog,epilog}-start` events and no corresponding `{prolog,epilog}-finish` event though..